tests/test_communication_utils.cpp

```diff
@@ -0,0 +1,196 @@
+    helper = &Dune::MPIHelper::instance(m_argc, m_argv);
+#ifdef MPI_2
+    MPI_Comm_create_errhandler(MPI_err_handler, &handler);
+   MPI_Comm_set_errhandler(MPI_COMM_WORLD, handler);
+#else
+        MPI_Errhandler_create(MPI_err_handler, &handler);
+        MPI_Errhandler_set(MPI_COMM_WORLD, handler);
+#endif
+#endif
+    }
+    ~MPIFixture()
+    {
+#if HAVE_MPI
+        MPI_Finalize();
+#endif
+    }
+    Dune::MPIHelper* helper;
+#if HAVE_MPI
+    MPI_Errhandler handler;
+#endif
+};
+
+BOOST_GLOBAL_FIXTURE(MPIFixture);
+
+template<class T, class C>
+std::tuple<std::vector<T>, std::vector<T>, std::vector<int>, std::vector<int>>
+createParallelData(const C& comm, T)
+{
+    std::vector<int> sizes(comm.size());
+    std::vector<int> displ(comm.size()+1);
+    std::vector<T> allVals;
+    
+    if (comm.rank() == 0)
+    {
+        // Initialize with random numbers.
+        std::random_device rndDevice;
+        std::mt19937 mersenneEngine {rndDevice()};  // Generates random integers
+        std::uniform_int_distribution<int> dist {1, 10};
+        
+        auto gen = [&dist, &mersenneEngine](){
+                       return dist(mersenneEngine);
+                   };
+        auto begin = std::begin(sizes);
+        auto end = std::end(sizes);
+        std::generate(begin, end, gen);
+        std::partial_sum(begin, end, std::begin(displ)+1);
+        allVals.resize(displ.back());
+        std::generate(std::begin(allVals), std::end(allVals), gen);
+    }
+    comm.broadcast(sizes.data(), sizes.size(), 0);
+    comm.broadcast(displ.data(), displ.size(), 0);
+    allVals.resize(displ.back()); // does nit change anything on 0
+    comm.broadcast(allVals.data(), allVals.size(), 0);
+    std::vector<T> myVals;
+    myVals.reserve(sizes[comm.rank()]);
+    auto beginAll = std::begin(allVals);
+
+    for ( auto val = beginAll + displ[comm.rank()]; val != beginAll + displ[comm.rank()+1];
+          ++val)
+    {
+        myVals.push_back(*val);
+    }
+
+    return { allVals, myVals, sizes, displ };
+}
+
+template<class T>
+void checkGlobalData(const std::vector<T>& data, const std::vector<T>& expected,
+                     const std::vector<int> displ, const std::vector<int> expectedDispl){
+    using std::begin;
+    using std::end;
+    BOOST_CHECK_EQUAL_COLLECTIONS(begin(data), end(data), begin(expected), end(expected));
+    BOOST_CHECK_EQUAL_COLLECTIONS(begin(displ), end(displ), begin(expectedDispl), end(expectedDispl));
+}
+
+template<class C>
+void testAllGatherv(const C& comm)
+{
+    //std::vector<int> sizes(comm.size());
+    std::vector<int> expectedDispl(comm.size()+1);
+    std::vector<int> displ;
+    std::vector<double> expectedAllVals;
+    std::vector<double> allVals;
+    std::vector<double> myVals;
+    double d=0;
+    
+    std::tie(expectedAllVals, myVals, std::ignore, expectedDispl) = createParallelData(comm, d);
+    std::tie(allVals, displ) = Opm::allGatherv(myVals, comm);
```

Comment:
Using `std::tie` means you need to
```C++
#include <tuple>
```
in this translation unit.

Reply:
Shouldn't the include of tuple in opm/grid/common/CommunicationUtils.hpp be enough?

Reply:
added them nevertheless to speed up things.

Reply:
Well, at least now it is